File: scripts/check-no-comments.mjs
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, extname } from "node:path";
import { spawnSync } from "node:child_process";
import ts from "typescript";

const EXCLUDED_DIRS = new Set([
  ".venv",
  "node_modules",
  ".next",
  ".git",
  "__pycache__",
  "playwright-report",
  "test-results",
  ".scratch",
]);

const TS_EXTENSIONS = new Set([".ts", ".tsx"]);
const EXCLUDED_FILES = new Set(["next-env.d.ts"]);

function collectFiles(root, extensions) {
  const results = [];
  const walk = (dir) => {
    for (const entry of readdirSync(dir)) {
      if (EXCLUDED_DIRS.has(entry) || EXCLUDED_FILES.has(entry)) continue;
      const fullPath = join(dir, entry);
      const stats = statSync(fullPath);
      if (stats.isDirectory()) {
        walk(fullPath);
      } else if (extensions.has(extname(entry))) {
        results.push(fullPath);
      }
    }
  };
  walk(root);
  return results;
}

function findCommentViolations(filePath) {
  const source = readFileSync(filePath, "utf-8");
  const sourceFile = ts.createSourceFile(
    filePath,
    source,
    ts.ScriptTarget.Latest,
    true,
    filePath.endsWith(".tsx") ? ts.ScriptKind.TSX : ts.ScriptKind.TS
  );
  const commentPositions = new Set();

  const visit = (node) => {
    if (node.kind === ts.SyntaxKind.JsxText) {
      return;
    }
    const ranges = [
      ...(ts.getLeadingCommentRanges(source, node.pos) ?? []),
      ...(ts.getTrailingCommentRanges(source, node.pos) ?? []),
    ];
    for (const range of ranges) {
      commentPositions.add(range.pos);
    }
    for (const child of node.getChildren(sourceFile)) {
      visit(child);
    }
  };
  visit(sourceFile);

  return [...commentPositions].map((position) => {
    const { line } = sourceFile.getLineAndCharacterOfPosition(position);
    return `${filePath}:${line + 1}: comment found`;
  });
}

function checkTypeScriptFiles(root) {
  const files = collectFiles(root, TS_EXTENSIONS);
  return files.flatMap(findCommentViolations);
}

function checkPythonFiles(root) {
  const venvPython =
    process.platform === "win32"
      ? join(".venv", "Scripts", "python.exe")
      : join(".venv", "bin", "python");
  const pythonCommand = statSync(venvPython, { throwIfNoEntry: false }) ? venvPython : "python3";
  const result = spawnSync(pythonCommand, ["scripts/check_no_comments.py", root], {
    encoding: "utf-8",
  });
  if (result.status !== 0) {
    return (result.stdout || result.stderr || "Python comment check failed.")
      .split("\n")
      .filter((line) => line.includes(": comment found") || line.includes(": docstring found"));
  }
  return [];
}

const root = process.cwd();
const violations = [...checkTypeScriptFiles(root), ...checkPythonFiles(root)];

if (violations.length > 0) {
  console.error(violations.join("\n"));
  console.error(`\n${violations.length} comment violation(s) found.`);
  process.exit(1);
}

console.log("No comments found in TypeScript or Python source files.");

File: scripts/check_no_comments.py
import ast
import io
import sys
import tokenize
from pathlib import Path

EXCLUDED_DIRS = {".venv", "node_modules", ".next", ".git", "__pycache__"}


def iter_python_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for path in root.rglob("*.py"):
        if any(part in EXCLUDED_DIRS for part in path.parts):
            continue
        files.append(path)
    return files


def find_comment_lines(source: str) -> list[int]:
    lines: list[int] = []
    tokens = tokenize.generate_tokens(io.StringIO(source).readline)
    for token in tokens:
        if token.type == tokenize.COMMENT:
            lines.append(token.start[0])
    return lines


def find_docstring_lines(source: str, path: Path) -> list[int]:
    tree = ast.parse(source, filename=str(path))
    lines: list[int] = []

    def check_body(body: list[ast.stmt]) -> None:
        if not body:
            return
        first = body[0]
        if (
            isinstance(first, ast.Expr)
            and isinstance(first.value, ast.Constant)
            and isinstance(first.value.value, str)
        ):
            lines.append(first.lineno)

    check_body(tree.body)
    for node in ast.walk(tree):
        if isinstance(node, ast.FunctionDef | ast.AsyncFunctionDef | ast.ClassDef):
            check_body(node.body)

    return lines


def main() -> int:
    root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path.cwd()
    violations: list[str] = []

    for path in iter_python_files(root):
        source = path.read_text(encoding="utf-8")
        for line in find_comment_lines(source):
            violations.append(f"{path}:{line}: comment found")
        for line in find_docstring_lines(source, path):
            violations.append(f"{path}:{line}: docstring found")

    if violations:
        print("\n".join(sorted(violations)))
        print(f"\n{len(violations)} comment/docstring violation(s) found.")
        return 1

    print("No comments or docstrings found in Python files.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
